packages/tailwindcss/src/variants.ts

```diff
@@ -1198,3 +1203,30 @@ export function substituteAtSlot(ast: AstNode[], nodes: AstNode[]) {
     }
   })
 }
+
+export function substituteAtVariant(ast: AstNode[], designSystem: DesignSystem): Features {
+  let features = Features.None
+  walk(ast, (variantNode, { replaceWith }) => {
+    if (variantNode.kind !== 'at-rule' || variantNode.name !== '@variant') return
+
+    // Starting with the `&` rule node
+    let node = styleRule('&', variantNode.nodes)
+
+    let variant = variantNode.params
+
+    let variantAst = designSystem.parseVariant(variant)
+    if (variantAst === null) {
+      throw new Error(`Cannot use \`@variant\` with unknown variant: ${variant}`)
+    }
+
+    let result = applyVariant(node, variantAst, designSystem.variants)
+    if (result === null) {
+      throw new Error(`Cannot use \`@variant\` with variant: ${variant}`)
+    }
+
+    // Update the variant at-rule node, to be the `&` rule node
+    replaceWith(node)
```

Comment:
If we never mutate the selector of the `&` node, then we could skip some layers by using:

```suggestion
    replaceWith(node.nodes)
```

But the final result should be the same. This is also the same behavior we had before so I think we can keep it like this.